### ISSUE TITLE
reportedStateCallback callback for IoTHubDeviceClient_LL_SendReportedState() sometimes is not called

### ISSUE THREAD

**Issue by vladimir-didenko** [Azure/azure-iot-sdk-c#652]
- **OS and version used:** MIPS Linux OpenWrt 3.18.29

- **SDK version used:** IoT Hub SDK for C 1.2.8 (2018-07-11)

# Description of the issue:
Sometimes reportedStateCallback callback for IoTHubDeviceClient_LL_SendReportedState() is not called at all. It looks that this issue somehow related to unstable network connection because the easiest way to reproduce this issue on our side is following procedure:
1. Re-connect our board to a different Wi-Fi access point
2. During this re-connection call IoTHubDeviceClient_LL_SendReportedState()
```
   twin_props_str = json_dumps(properties, JSON_COMPACT);
   ASSERT(NULL != twin_props_str);
   twin_props_length = strlen(twin_props_str);

   iot_sdk_ret = IoTHubDeviceClient_LL_SendReportedState(azure_ctx->azct_client,
                                                                                          (const unsigned char *)twin_props_str,
                                                                                          twin_props_length,
                                                                                          s_azure_send_reported_twin_props_cb,
                                                                                          azure_ctx);
   ASSERT(IOTHUB_CLIENT_OK == iot_sdk_ret);
```
3. s_azure_send_reported_twin_props_cb() is never called

Note 1: it is not 100% in every environment e.g. we could quite easily to reproduce it in one lab and can't reproduce it in another. 

Note 2: provided logs (see below) we made on build that calls IoTHubDeviceClient_LL_DoWork() once a second but we still could reproduce the issue with 100ms interval.

Note 3: connectionStatusCallback() callback for IoTHubDeviceClient_LL_SetConnectionStatusCallback() is called 2 times with IOTHUB_CLIENT_CONNECTION_AUTHENTICATED status. First time it is called during first connect (before access point switching), second time it is called after switching to a new access point. No connection errors are reported between this calls though I personally expect connection failure in case of access point switching.

# Code sample exhibiting the issue:
https://gist.github.com/vladimir-didenko/ef2672c38239adcf896456ccfd05b2b3#file-azure-c

# Console log of the issue:
Note: we are switching to a new access point at second 1013 (monotonic time).

Application log from azure module - https://gist.github.com/vladimir-didenko/ef2672c38239adcf896456ccfd05b2b3#file-device-log

Azure SDK log - https://gist.github.com/vladimir-didenko/ef2672c38239adcf896456ccfd05b2b3#file-azure_sdk-log

As you see at second 1014 SDK tries to send update for device twin
```
[  1014.169275135] -> 21:56:56 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/PATCH/properties/reported/?$rid=24 | PAYLOAD_LEN: 692
```

but doesn't receive confirmation and re-establishes connection
```
[  1044.198915346] -> 21:57:26 PINGREQ
[  1059.215722965] -> 21:57:41 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/000680015f3d/messages/events/name=common.diagnostics | PACKET_ID: 23 | PAYLOAD_LEN: 27
[  1060.217025155] ERR: iothubtransport_mqtt_common.c:1674 mqtt_error_callbackMqtt Ping Response was not encountered.  Reconnecting device...
[  1060.222630894] -> 21:57:42 DISCONNECT
[  1064.68034430] -> 21:57:46 CONNECT | VER: 4 | KEEPALIVE: 30 | FLAGS: 128 | USERNAME: somehost.azure-devices.net/000680015f3d/?api-version=2017-11-08-preview&DeviceClientType=iothubclient%2f1.2.8%20(native%3b%20Linux%3b%20mips) | CLEAN: 0
[  1065.69838727] <- 21:57:47 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
[  1066.72646255] -> 21:57:48 SUBSCRIBE | PACKET_ID: 25 | TOPIC_NAME: devices/000680015f3d/messages/devicebound/# | QOS: 1 | TOPIC_NAME: $iothub/twin/res/# | QOS: 0 | TOPIC_NAME: $iothub/twin/PATCH/properties/desired/# | QOS: 0 | TOPIC_NAME: $iothub/methods/POST/# | QOS: 0
[  1067.74520113] <- 21:57:49 SUBACK | PACKET_ID: 25 | RETURN_CODE: 1 | RETURN_CODE: 0 | RETURN_CODE: 0 | RETURN_CODE: 0
[  1068.76710827] -> 21:57:50 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/GET/?$rid=26
[  1069.79218820] <- 21:57:51 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE = 0x00 | TOPIC_NAME: $iothub/twin/res/200/?$rid=26 | PAYLOAD_LEN: 929
[  1098.110967912] -> 21:58:20 PINGREQ
[  1099.112802188] <- 21:58:21 PINGRESP
[  1120.140674329] -> 21:58:42 DISCONNECT
[  1124.267756201] -> 21:58:46 CONNECT | VER: 4 | KEEPALIVE: 30 | FLAGS: 128 | USERNAME: somehost.azure-devices.net/000680015f3d/?api-version=2017-11-08-preview&DeviceClientType=iothubclient%2f1.2.8%20(native%3b%20Linux%3b%20mips) | CLEAN: 0
[  1125.269590430] <- 21:58:47 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
```
At second 1125 SDK calls connection status callback but doesn't call callback for device twin (didn't do it before this moment and will not do it in future).

**Comment by vladimir-didenko** [comment]
I don't know if it is helpful but there is additional information. I made fix for our code that re-sends device twin every time connection status callback (IOTHUB_CLIENT_CONNECTION_STATUS_CALLBACK) is called and it doesn't help.

**Comment by massand** [comment]
@vladimir-didenko Thanks for the comprehensive logs and sample! We will investigate the logs and repro this. I just want to clarify - this isn't a 100% repro on your end and when it does repro, you don't receive any more ACKs for reported state changes? Does the IoT Hub portal show the twin update? Thanks in advance!

**Comment by vladimir-didenko** [comment]
@massand 
> this isn't a 100% repro on your end

Right. Specific network conditions are required (e.g. switch access point).

>  when it does repro, you don't receive any more ACKs for reported state changes?

Right.

> Does the IoT Hub portal show the twin update?

No it doesn't.

**Comment by massand** [comment]
@vladimir-didenko  Sorry for the delay in responding. I wanted to validate few scenarios and figure out the best approach to tackle this in parallel. 

Here is what I found:
1) Once a disconnection occurs, the 'reported state pending ack from service' queue doesn't exist in the SDK yet (unlike messages which have an 'message event list waiting for pending ack' concept, where the SDK has a callback queued somewhere). So, the fact that we don't ack back on reported state sends and drop them on the floor on disconnection is by design, unfortunately. However, we are currently working on enhancing the SDK to make the device twin reported state expected behavior (ack back after connection is restored) consistent with the messages behavior. Stay tuned on this.

2) If I use the connection status callback approach to resubmit the reported state change, my report state callback WILL fire properly for subsequent reported state changes. I cannot reproduce the behavior you observed about reported status callback not firing after connection is re-established and new reported state changes are fired. 

[sample_tracelog.txt](https://github.com/Azure/azure-iot-sdk-c/files/2533120/sample_tracelog.txt)
[devicetwin_simplesample_c.txt](https://github.com/Azure/azure-iot-sdk-c/files/2533148/devicetwin_simplesample_c.txt)

I have modified the current devicetwin_simplesample.c for your convenience and attached the logs I observed. To make the repro easy on your end, I recommend the following:
1) put a breakpoint at iothub_client_core_ll.c, line 2017 
2) kill your network connection after you observe the reported state has been sent to the service 
3) re-establish your network connection after the transport reports connection is dropped and observe the reported state callback

**Comment by vladimir-didenko** [comment]
@massand 
Thank you! We will try to add suggested change and re-test behavior on our side. There is one note though - we already tried this workaround previously but with older version (1.28). And it didn't help. Anyway we will re-check if it works with latest SDK version.
